### ISSUE TITLE
Id Modules: not respecting publisher configuration

### ISSUE THREAD

**Issue by patmmccann** [prebid/Prebid.js#10710]
## Type of issue
Bug

## Description
![image](https://github.com/prebid/Prebid.js/assets/1683175/e354cebf-b874-4175-9727-41eefda12e35)
shows publishers can choose where to set storage, but several modules ignore

https://github.com/criteo-forks/Prebid.js/blob/a22fee00f1ab9084b297ec6d25297edef02382d1/modules/criteoIdSystem.js#L39
https://github.com/prebid/Prebid.js/blob/015f12b3a2ab62d993e2578d6725b6acf8456910/modules/connectIdSystem.js#L45

Lotame appears to not have this setting available and just sets both
https://github.com/prebid/Prebid.js/blob/015f12b3a2ab62d993e2578d6725b6acf8456910/modules/lotamePanoramaIdSystem.js#L45

LiveRamp appears to not use the parameter for html5 or cookie but instead has an additional proprietary parameter that seems related

https://github.com/prebid/Prebid.js/blob/015f12b3a2ab62d993e2578d6725b6acf8456910/modules/identityLinkIdSystem.js#L122

![image](https://github.com/prebid/Prebid.js/assets/1683175/cd9b17fb-e7c1-46be-810c-79b47cbacf8c)

**Comment by dgirardi** [comment]
These are ID modules that use storage directly (grepping for `getStorageManager`). It doesn't necessarily mean that they don't respect publisher config, but in theory ID modules do not need to deal with storage - it's done for them by the base ID module.

```
        admixerIdSystem.js  
        adqueryIdSystem.js  
        adriverIdSystem.js  
        amxIdSystem.js  
        connectIdSystem.js  
        criteoIdSystem.js  
        czechAdIdSystem.js  
        dacIdSystem.js  
        deepintentDpesIdSystem.js  
        euidIdSystem.js  
        freepassIdSystem.js  
        ftrackIdSystem.js  
        gravitoIdSystem.js  
        growthCodeIdSystem.js  
        hadronIdSystem.js  
        id5IdSystem.js  
        identityLinkIdSystem.js  
        idxIdSystem.js  
        imuIdSystem.js  
        intentIqIdSystem.js  
        liveIntentIdSystem.js  
        lotamePanoramaIdSystem.js  
        merkleIdSystem.js  
        mwOpenLinkIdSystem.js  
        naveggIdSystem.js  
        novatiqIdSystem.js  
        pairIdSystem.js  
        parrableIdSystem.js  
        publinkIdSystem.js  
        quantcastIdSystem.js  
        teadsIdSystem.js  
        uid2IdSystem.js  
        zeotapIdPlusIdSystem.js  
```

**Comment by patmmccann** [comment]
We can see id modules, as noted in the example above, violating the publisher parameter choices for storage.type listed at `https://docs.prebid.org/dev-docs/modules/userId.html#basic-configuration`

For example, panoramasets the cookie when the publisher chooses html5. This means it is operating outside of how it is configured and documented to behave and is a bug we need to fix.

**Comment by patmmccann** [comment]
I propose we boot all id modules doing this in Prebid 9. @jdwieland8282 let's talk about it in committee?

**Comment by danielsao** [comment]
@patmmccann LiveRamp user id module respects the publisher settings in `storage.type`. The line of code you are referring to is about logging the envelope source type in a cookie for analytics purposes and not about overwriting the storage location.

**Comment by leonardlabat** [comment]
Hello @patmmccann 

> However, Criteo team can you also have your bidder module do the same when it sets cto_bundle?

Do you have an example showing how the storage configuration can be retrieved from a bidder adapter?

**Comment by patmmccann** [comment]
Might be too ambitious for prebid 9? what do you think @lcorrigall @dgirardi

**Comment by patmmccann** [comment]
> Do you have an example showing how the storage configuration can be retrieved from a bidder adapter?

You could take it as a bidder param potentially?

**Comment by jdwieland8282** [comment]
I don't think we should boot the modules who are out of compliance.  If we include it in pb 9 what does that practically mean?

**Comment by muuki88** [comment]
If being removed in pb9 , publishers that use those id modules would not upgrade, until the adapter has been resubmitted.

I'm actually in favor of removing ( booting? ) the modules that do not comply. We had issues with cookie headers being too large and when the storage settings are ignored, this may cause frustration.

**Comment by jdwieland8282** [comment]
Thats a fair point @muuki88 

@dgirardi I see your list above, how can I figure out which of those do not respect the pubs `cookies|html5` choice?

**Comment by patmmccann** [comment]
Committee discussion result: we should subset the above list to those that are writing outside of the configured location, not reading outside of configuration.

**Comment by jdwieland8282** [comment]
Let me know how I can help compiling the list.

**Comment by patmmccann** [comment]
Checking the above list for writes:

https://github.com/prebid/Prebid.js/blob/dbf981739da6d402b0c09e0aa2289cdbb6e86be1/modules/adqueryIdSystem.js#L114
https://github.com/prebid/Prebid.js/blob/dbf981739da6d402b0c09e0aa2289cdbb6e86be1/modules/adriverIdSystem.js#L73
https://github.com/prebid/Prebid.js/blob/dbf981739da6d402b0c09e0aa2289cdbb6e86be1/modules/connectIdSystem.js#L58
https://github.com/prebid/Prebid.js/blob/dbf981739da6d402b0c09e0aa2289cdbb6e86be1/modules/dacIdSystem.js#L94

This appears to be a bug, as 'both' is now an option
https://github.com/prebid/Prebid.js/blob/dbf981739da6d402b0c09e0aa2289cdbb6e86be1/modules/uid2IdSystem_shared.js#L685

**Comment by jdwieland8282** [comment]
@dgirardi can you find a module that respects the `param.storage.type` so we id maintainers can see how this should be supported?

**Comment by patmmccann** [comment]
> [@dgirardi](https://github.com/dgirardi) can you find a module that respects the `param.storage.type` so we id maintainers can see how this should be supported?

Yandex recently simplified their submission to conform

**Comment by abazylewicz-id5** [comment]
Is it ok if in our user module, when we do a server call to get an ID, in the response we set cookie with given ID? Or we would need to require cookie storage/not set the cookie when our module is configured to use only `html5` storage type?

If we have an option to use external module through [explicit module configuration parameter](https://docs.prebid.org/dev-docs/modules/userid-submodules/id5.html#id5-id-configuration), does the external module also should respect prebid storage configuration or it's enough if we add to the description that the external module uses some additional storage regardless of what's set in prebid storage type?

**Comment by patmmccann** [comment]
Your response set cookie header is no problem. The problem is if your js sets a cookie in the first party context when the publisher directed you to use local storage for the first party

**Comment by abazylewicz-id5** [comment]
Thanks @patmmccann , for some reason I missed the fact that Prebid indeed can set first party cookies. We do not store any first party cookies in external module, and submitted the fix to the storing on prebid side in this MR: #11965

**Comment by jdwieland8282** [comment]
I heard offline from Mediawalla and Lotame that they have fixed their modules, can anyone confirm?

**Comment by jdwieland8282** [comment]
Mediawallah's fix https://github.com/prebid/Prebid.js/pull/12149

**Comment by Tonsil** [comment]
> I heard offline from Mediawalla and Lotame that they have fixed their modules, can anyone confirm?

As a Lotame representative, I can tell you that work has been prioritized and planned for Q4 but not yet completed. Is there a deadline for when this change needs to go live? I tried to find a schedule posted for "10.0 API Change" timing, but was unsuccessful.

**Comment by jdwieland8282** [comment]
Can we generate an updated list of offending ID modules? Christian and I will begin outreach again.

**Comment by patmmccann** [comment]
Instead of kicking out any adapters, we can make the id submodule call to set a storage contingent on not violating existing publisher configuration for that storage type. The choice of key is out of scope but may also be something people wish to address in anopther issue.

**Comment by patmmccann** [comment]
The cleanest violation I am aware of is https://github.com/prebid/Prebid.js/blob/dbf981739da6d402b0c09e0aa2289cdbb6e86be1/modules/teadsIdSystem.js#L79

**Comment by jdwieland8282** [comment]
The Identity PMC agrees with this approach 

> Instead of kicking out any adapters, we can make the id submodule call to set a storage contingent on not violating existing publisher configuration for that storage type. The choice of key is out of scope but may also be something people wish to address in another issue.

To mitigate as much as possible any unintended effects on publishers or id maintainers we'd as that:
1. this change be included in the PB 10 release notes, specifically we should name the id modules that will be impacted
2. console log an error when a `write` is blocked
3. console log a warning when a `write` puts the 'name' or 'expiration' in a location not specified by the publisher in the 'params'
